ARMmbed/mbedtls
# Add parameter validation framework

## Description
This PR introduces, MBEDTLS_CHECK_PARAMS, a configuration symbol to enable compilation of parameter validation checks into the the build.

Additional information is available in `config.h`, in the documenting comment next to the introduced define.

The checks in the build are demonstrated in aes.c.

This PR is provided for evaluation of the means of validating parameters, and is not yet functionally complete.

## Status
**READY**

## Requires Backporting
NO (unless new requirements come up)

## Migrations
NO

## Additional comments
This is the leading change of a train of commits unifying the function argument validation in the library.

## Todos

## Manual testing

The following manual testing was done by Manuel - please double-check if other things should be tested manually:

- put `if( ctx == NULL ) return;` in front of `AES_VALIDATE( ctx != NULL )` in `mbedtls_aes_init()` and observe that the test suite fails as it should
- put `AES_VALIDATE( ctx != NULL )` in front of `if( ctx == NULL ) return;` in `mbedtls_aes_free()` and observe that test test suite fails as it should
- put `if( ctx == NULL ) return( MBEDTLS_ERR_AES_BAD_INPUT_DATA );` in front of `AES_VALIDATE_RET( ctx != NULL )` in `mbedtls_aes_setkey_end()` and observe that the test suite fails as it should
- change `MBEDTLS_ERR_AES_BAD_INPUT_DATA` to some other value in the definition of `AES_VALIDATE_RET()` and observe that the test suite fails as it should
- run `make '-Werror -O1' all check` with a full config and observe that it passes
- run the two paragraphs added to `all.sh` and observe that they pass
